### PR TITLE
OAuth Rework

### LINKED ISSUE #530: OAuth login fails to verify after restart
```
00:00:02.160 [ERROR] Unhandled exception|System.Net.Http.HttpRequestException: Unauthorized
   at NexusMods.Networking.NexusWebApi.Client.SendAsync[T](HttpRequestMessage message, JsonTypeInfo`1 typeInfo, CancellationToken token) in /ssd-pool/projects/NexusMods/app/main/src/Networking/NexusMods.Networking.NexusWebApi/Client.cs:line 172
   at NexusMods.Networking.NexusWebApi.Client.SendAsync[T](HttpRequestMessage message, JsonTypeInfo`1 typeInfo, CancellationToken token) in /ssd-pool/projects/NexusMods/app/main/src/Networking/NexusMods.Networking.NexusWebApi/Client.cs:line 190
   at NexusMods.Networking.NexusWebApi.Client.SendAsync[T](HttpRequestMessage message, CancellationToken token) in /ssd-pool/projects/NexusMods/app/main/src/Networking/NexusMods.Networking.NexusWebApi/Client.cs:line 156
   at NexusMods.Networking.NexusWebApi.Client.ModFilesAsync(String domain, ModId modId, CancellationToken token) in /ssd-pool/projects/NexusMods/app/main/src/Networking/NexusMods.Networking.NexusWebApi/Client.cs:line 150
   at NexusMods.Networking.NexusWebApi.NMA.Extensions.ClientExtensions.ModFilesAsync(Client client, GameDomain domain, ModId modId, CancellationToken token) in /ssd-pool/projects/NexusMods/app/main/src/Networking/NexusMods.Networking.NexusWebApi.NMA/Extensions/ClientExtensions.cs:line 67
   at NexusMods.Networking.NexusWebApi.NMA.OAuth2MessageFactory.Verify(Client client, CancellationToken cancel) in /ssd-pool/projects/NexusMods/app/main/src/Networking/NexusMods.Networking.NexusWebApi.NMA/OAuth2MessageFactory.cs:line 55
   at NexusMods.Networking.NexusWebApi.NMA.LoginManager.Verify() in /ssd-pool/projects/NexusMods/app/main/src/Networking/NexusMods.Networking.NexusWebApi.NMA/LoginManager.cs:line 64
   at NexusMods.Networking.NexusWebApi.NMA.LoginManager.<.ctor>b__14_2(Boolean _) in /ssd-pool/projects/NexusMods/app/main/src/Networking/NexusMods.Networking.NexusWebApi.NMA/LoginManager.cs:line 58
```

Steps to reproduce:

1) Login
2) Close the App
3) Open the App again

The Login button is not clickable, the JWT Token is still in the Datastore and the Login Manager keeps throwing unhandled exceptions.

### PR DESCRIPTION
**Pull request by erri120**
Resolves #530 and #525.

Changes:

- We now have our own OAuth client instead of the old Vortex client.
- Updated OAuth scopes that fixed some endpoints.
- The OAuth token is now cached and automatically refreshes itself
- UserInfo is now cached in the login manager with a lifetime of 10 mins.
- Fixed various bugs.

### CODE REVIEW

File: src/Networking/NexusMods.Networking.NexusWebApi.NMA/ApiKeyMessageFactory.cs [hunk 1]
```diff
@@ -75,7 +75,7 @@ public ValueTask SetApiKey(string apiKey)
         {
             Name = result.Data.Name,
             IsPremium = result.Data.IsPremium,
-            IsSupporter = result.Data.IsSupporter,
```
Comment (Al12rs) [inline src/Networking/NexusMods.Networking.NexusWebApi.NMA/ApiKeyMessageFactory.cs:78]: So I'm guessing that IsSupporter data isn't available in the response?

Comment (erri120) [reply src/Networking/NexusMods.Networking.NexusWebApi.NMA/ApiKeyMessageFactory.cs:78]: The "supporter" status is deprecated.

File: src/Networking/NexusMods.Networking.NexusWebApi.NMA/LoginManager.cs [hunk 2]
```diff
@@ -44,25 +48,45 @@ public class LoginManager
...
             .Where(id => id.Equals(JWTTokenEntity.StoreId))
-            .Select(_ => true)
-            .StartWith(true)
-            .SelectMany(async _ => await Verify());
+            .ObserveOn(TaskPoolScheduler.Default)
+            .SelectMany(async _ => await Verify(CancellationToken.None));
     }
 
-    private async Task<UserInfo?> Verify()
+    private CachedObject<UserInfo> _cachedUserInfo = new(TimeSpan.FromHours(1));
```
[the remark this answers is not in the document]
Comment (erri120) [reply src/Networking/NexusMods.Networking.NexusWebApi.NMA/LoginManager.cs:71]: Doesn't really make a difference:

```csharp
private static readonly TimeSpan UserInfoCacheLifespan = TimeSpan.FromHours(1);
private CachedObject<UserInfo> _cachedUserInfo = new(UserInfoCacheLifespan);
```

[the remark this answers is not in the document]
Comment (Al12rs) [reply src/Networking/NexusMods.Networking.NexusWebApi.NMA/LoginManager.cs:71]: I had a look and that is the only explicit lifetime I found in the PR, thought there might have been more. If that is correct then we can leave this one

File: src/Networking/NexusMods.Networking.NexusWebApi.NMA/OAuth.cs [hunk 3]
```diff
@@ -48,39 +52,38 @@ public class OAuth
...
-    /// make an authorization request
+    /// Make an authorization request
     /// </summary>
-    /// <param name="cancel"></param>
+    /// <param name="cancellationToken"></param>
     /// <returns>task with the jwt token once we receive one</returns>
-    public async Task<JwtTokenReply> AuthorizeRequest(CancellationToken cancel)
+    public async Task<JwtTokenReply> AuthorizeRequest(CancellationToken cancellationToken)
     {
-        _logger.LogInformation("Starting NexusMods OAuth2 authorization request");
```
Comment (Al12rs) [inline src/Networking/NexusMods.Networking.NexusWebApi.NMA/OAuth.cs:57]: Why remove logging?

File: src/Networking/NexusMods.Networking.NexusWebApi.NMA/OAuth.cs [hunk 4]
```diff
@@ -127,37 +130,42 @@ private async Task<JwtTokenReply> AuthorizeToken(string verifier, string code, C
         return JsonSerializer.Deserialize<JwtTokenReply>(responseString);
     }
 
-    private string SanitizeBase64(string input)
-    {
-        return input
-            .Replace("+", "-")
-            .Replace("/", "_")
-            .TrimEnd('=');
-    }
```
[the remark this answers is not in the document]
Comment (erri120) [reply src/Networking/NexusMods.Networking.NexusWebApi.NMA/OAuth.cs:136]: Most of the code in this file was directly translated from the Vortex implementation to C# without making use of .NET specific methods and libraries.

[the remark this answers is not in the document]
Comment (Al12rs) [reply src/Networking/NexusMods.Networking.NexusWebApi.NMA/OAuth.cs:136]: Yeah, reading some more of the code I see you used Encode and stuff later, which I'm guessing handles this stuff

**Review by Al12rs — APPROVED** [review]

**Pull request merged**